scenarios/helpers/iq_metrics.erl

```diff
@@ -0,0 +1,108 @@
+-module(iq_metrics).
+
+-export([start/1, request/2, request/3, timeout/2, response/2, new_id/1]).
+
+-record(request, {
+    key :: {Id :: binary(), pid()},
+    name :: name(),
+    time :: integer(),
+    tref :: timer:tref() | undefined | timed_out
+}).
+
+-type name() :: atom().
+
+-define(TABLE, request).
+
+-spec start([name()]) -> ok.
+start(Names) ->
+    {ok, App} = application:get_application(),
+    PrivDir = code:priv_dir(App),
+    application:set_env(mnesia, dir, PrivDir),
+
+    mnesia:create_schema([node()]),
+    application:ensure_started(mnesia),
+    mnesia:create_table(?TABLE, [{attributes, record_info(fields, request)}]),
+
+    Suffixes = [[request], [response], [timeout], [response, result], [response, error], [timeout, result], [timeout, error]],
+    [begin
+         [amoc_metrics:init(counters, [Name | Suffix]) || Suffix <- Suffixes],
+         amoc_metrics:init(times, [Name, response, time])
+     end || Name <- Names],
+    ok.
+
+%% API
+
+-spec request(name(), binary()) -> any().
+request(Name, Id) ->
+    process_request(Name, Id, undefined).
+
+-spec request(name(), binary(), timeout()) -> any().
+request(Name, Id, Timeout) ->
+    {ok, TRef} = timer:apply_after(Timeout, ?MODULE, timeout, [{Id, self()}, no_delete]),
+    process_request(Name, Id, TRef).
+
+-spec timeout({binary(), pid()}, delete | no_delete) -> any();
+             (binary(), delete | no_delete) -> any().
+timeout({_Id, _Pid} = Key, no_delete) ->
+    F = fun() ->
+            case mnesia:read(?TABLE, Key) of
+                [Req] -> mnesia:write(Req#request{tref = timed_out}),
+                         Req;
+                [] -> ok % timed out during success or failure execution
+            end
+        end,
+    case mnesia:transaction(F) of
```

Comment:
Do we need transaction here? Would `mnesia:dirty_read` be enough here?

Reply:
As I understand it, we need a transaction, as it may happen that an amoc user reports a response (which results in deleting a record from the table) in between reading and writing here.